From 7e4535b2daf9f54904fd86fa15e6937a04a3281d Mon Sep 17 00:00:00 2001
From: liangshuang <liangshuang@eswincomputing.com>
Date: Mon, 9 Dec 2024 17:50:25 +0800
Subject: [PATCH 333/481] WIN2030-16760:feat:adapt eth driver to board z530.

Changelogs:
1. adapt eth driver to board z530.
2. modify the eth driver to adapt to the delay parameters of different boards.
3. remove iommu in eth node.

Change-Id: I5bef1e104bf4aa9158d0289baa49d5b3d1b313b0
Signed-off-by: liangshuang <liangshuang@eswincomputing.com>
---
 arch/riscv/boot/dts/eswin/eic7700-evb-a2.dts  |  8 +-
 arch/riscv/boot/dts/eswin/eic7700-evb.dts     |  8 +-
 .../dts/eswin/eic7700-hifive-premier-p550.dts |  6 ++
 .../boot/dts/eswin/eic7700-som260-a1.dtsi     |  2 -
 arch/riscv/boot/dts/eswin/eic7700-z530.dts    | 13 +++
 .../boot/dts/eswin/eic7702-evb-a1-d0.dts      |  8 +-
 .../boot/dts/eswin/eic7702-evb-a1-d1.dts      |  8 +-
 arch/riscv/boot/dts/eswin/eic7702-evb-a1.dts  | 16 +++-
 .../dts/eswin/eswin-win2030-die0-soc.dtsi     |  2 +
 .../dts/eswin/eswin-win2030-die1-soc.dtsi     |  2 +
 .../ethernet/stmicro/stmmac/dwmac-win2030.c   | 87 ++++++++++++-------
 11 files changed, 116 insertions(+), 44 deletions(-)

diff --git a/arch/riscv/boot/dts/eswin/eic7700-evb-a2.dts b/arch/riscv/boot/dts/eswin/eic7700-evb-a2.dts
index 576aa7fa317b..42a2eb700568 100644
--- a/arch/riscv/boot/dts/eswin/eic7700-evb-a2.dts
+++ b/arch/riscv/boot/dts/eswin/eic7700-evb-a2.dts
@@ -508,7 +508,9 @@ &d0_gmac0 {
 	rst-gpios = <&portc 30 GPIO_ACTIVE_LOW>;
 	eswin,rgmiisel = <&pinctrl 0x290 0x3>;
 	eswin,led-cfgs = <0x6100 0xa40 0x420>;
-
+	dly-param-1000m = <0x23232323 0x800c8023 0x0c0c0c0c>;
+	dly-param-100m = <0x50505050 0x803f8050 0x3f3f3f3f>;
+	dly-param-10m = <0 0 0>;
 	status = "okay";
 };
 
@@ -518,7 +520,9 @@ &d0_gmac1 {
 	rst-gpios = <&porta 16 GPIO_ACTIVE_LOW>;
 	eswin,rgmiisel = <&pinctrl 0x294 0x3>;
 	eswin,led-cfgs = <0x6100 0xa40 0x420>;
-
+	dly-param-1000m = <0x25252525 0x80268025 0x26262626>;
+	dly-param-100m = <0x48484848 0x80588048 0x58585858>;
+	dly-param-10m = <0 0 0>;
 	status = "okay";
 };
 
diff --git a/arch/riscv/boot/dts/eswin/eic7700-evb.dts b/arch/riscv/boot/dts/eswin/eic7700-evb.dts
index 99068867bd15..64fac2c5906c 100644
--- a/arch/riscv/boot/dts/eswin/eic7700-evb.dts
+++ b/arch/riscv/boot/dts/eswin/eic7700-evb.dts
@@ -467,7 +467,9 @@ &d0_gmac0 {
 	rst-gpios = <&portc 30 GPIO_ACTIVE_LOW>;
 	eswin,rgmiisel = <&pinctrl 0x290 0x3>;
 	eswin,led-cfgs = <0x6251 0x6251 0x6251>;
-
+	dly-param-1000m = <0x23232323 0x800c8023 0x0c0c0c0c>;
+	dly-param-100m = <0x50505050 0x803f8050 0x3f3f3f3f>;
+	dly-param-10m = <0 0 0>;
 	status = "okay";
 };
 
@@ -477,7 +479,9 @@ &d0_gmac1 {
 	rst-gpios = <&porta 16 GPIO_ACTIVE_LOW>;
 	eswin,rgmiisel = <&pinctrl 0x294 0x3>;
 	eswin,led-cfgs = <0x6251 0x6251 0x6251>;
-
+	dly-param-1000m = <0x25252525 0x80268025 0x26262626>;
+	dly-param-100m = <0x48484848 0x80588048 0x58585858>;
+	dly-param-10m = <0 0 0>;
 	status = "okay";
 };
 
diff --git a/arch/riscv/boot/dts/eswin/eic7700-hifive-premier-p550.dts b/arch/riscv/boot/dts/eswin/eic7700-hifive-premier-p550.dts
index b19e3e67dd7a..66f5621a146f 100644
--- a/arch/riscv/boot/dts/eswin/eic7700-hifive-premier-p550.dts
+++ b/arch/riscv/boot/dts/eswin/eic7700-hifive-premier-p550.dts
@@ -457,6 +457,9 @@ &d0_gmac0 {
 	rst-gpios = <&portd 10 GPIO_ACTIVE_LOW>;
 	eswin,rgmiisel = <&pinctrl 0x290 0x3>;
 	eswin,led-cfgs = <0x6100 0xa40 0x420>;
+	dly-param-1000m = <0x23232323 0x800c8023 0x0c0c0c0c>;
+	dly-param-100m = <0x50505050 0x803f8050 0x3f3f3f3f>;
+	dly-param-10m = <0 0 0>;
 };
 
 &d0_gmac1 {
@@ -466,6 +469,9 @@ &d0_gmac1 {
 	rst-gpios = <&portd 15 GPIO_ACTIVE_LOW>;
 	eswin,rgmiisel = <&pinctrl 0x294 0x3>;
 	eswin,led-cfgs = <0x6100 0xa40 0x420>;
+	dly-param-1000m = <0x25252525 0x80268025 0x26262626>;
+	dly-param-100m = <0x48484848 0x80588048 0x58585858>;
+	dly-param-10m = <0 0 0>;
 };
 
 &d0_sata {
diff --git a/arch/riscv/boot/dts/eswin/eic7700-som260-a1.dtsi b/arch/riscv/boot/dts/eswin/eic7700-som260-a1.dtsi
index 6ae44bf9b159..4f93ff6213e8 100644
--- a/arch/riscv/boot/dts/eswin/eic7700-som260-a1.dtsi
+++ b/arch/riscv/boot/dts/eswin/eic7700-som260-a1.dtsi
@@ -453,7 +453,6 @@ &d0_gmac0 {
 	pinctrl-names = "default";
 	eswin,rgmiisel = <&pinctrl 0x290 0x3>;
 	eswin,led-cfgs = <0x6100 0xa40 0x420>;
-
 	status = "okay";
 };
 
@@ -461,7 +460,6 @@ &d0_gmac1 {
 	pinctrl-names = "default";
 	eswin,rgmiisel = <&pinctrl 0x294 0x3>;
 	eswin,led-cfgs = <0x6100 0xa40 0x420>;
-
 	status = "okay";
 };
 
diff --git a/arch/riscv/boot/dts/eswin/eic7700-z530.dts b/arch/riscv/boot/dts/eswin/eic7700-z530.dts
index 0512aadbfb80..4059881343d1 100644
--- a/arch/riscv/boot/dts/eswin/eic7700-z530.dts
+++ b/arch/riscv/boot/dts/eswin/eic7700-z530.dts
@@ -38,6 +38,19 @@ ap_backup {
 	};
 };
 
+&d0_gmac0 {
+	dly-param-1000m = <0x23232323 0xa0209023 0x20202020>;
+	dly-param-100m = <0x50505050 0x803f8050 0x3f3f3f3f>;
+	dly-param-10m = <0 0 0>;
+};
+
+&d0_gmac1 {
+	dly-param-1000m = <0x00000000 0xe002e000 0x04040404>;
+	dly-param-100m = <0x48484848 0x80588048 0x58585858>;
+	dly-param-10m = <0 0 0>;
+};
+
+
 &pcie {
 	pinctrl-names = "default";
 	pinctrl-0 = <&pinctrl_gpio10_default>;
diff --git a/arch/riscv/boot/dts/eswin/eic7702-evb-a1-d0.dts b/arch/riscv/boot/dts/eswin/eic7702-evb-a1-d0.dts
index 7688d0071ba1..0dd1eaa522a4 100644
--- a/arch/riscv/boot/dts/eswin/eic7702-evb-a1-d0.dts
+++ b/arch/riscv/boot/dts/eswin/eic7702-evb-a1-d0.dts
@@ -429,7 +429,9 @@ &d0_gmac0 {
 	rst-gpios = <&porta 15 GPIO_ACTIVE_LOW>;
 	eswin,rgmiisel = <&pinctrl 0x290 0x3>;
 	eswin,led-cfgs = <0x6100 0xa40 0x420>;
-
+	dly-param-1000m = <0x23232323 0x800c8023 0x0c0c0c0c>;
+	dly-param-100m = <0x50505050 0x803f8050 0x3f3f3f3f>;
+	dly-param-10m = <0 0 0>;
 	status = "okay";
 };
 
@@ -439,7 +441,9 @@ &d0_gmac1 {
 	rst-gpios = <&porta 14 GPIO_ACTIVE_LOW>;
 	eswin,rgmiisel = <&pinctrl 0x294 0x3>;
 	eswin,led-cfgs = <0x6100 0xa40 0x420>;
-
+	dly-param-1000m = <0x25252525 0x80268025 0x26262626>;
+	dly-param-100m = <0x48484848 0x80588048 0x58585858>;
+	dly-param-10m = <0 0 0>;
 	status = "okay";
 };
 
diff --git a/arch/riscv/boot/dts/eswin/eic7702-evb-a1-d1.dts b/arch/riscv/boot/dts/eswin/eic7702-evb-a1-d1.dts
index 110e4eab9a6d..0b2d07df055c 100644
--- a/arch/riscv/boot/dts/eswin/eic7702-evb-a1-d1.dts
+++ b/arch/riscv/boot/dts/eswin/eic7702-evb-a1-d1.dts
@@ -443,7 +443,9 @@ &d0_gmac0 {
 	rst-gpios = <&porta 13 GPIO_ACTIVE_LOW>;
 	eswin,rgmiisel = <&pinctrl 0x290 0x3>;
 	eswin,led-cfgs = <0x6100 0xa40 0x420>;
-
+	dly-param-1000m = <0x23232323 0x800c8023 0x0c0c0c0c>;
+	dly-param-100m = <0x50505050 0x803f8050 0x3f3f3f3f>;
+	dly-param-10m = <0 0 0>;
 	status = "okay";
 };
 
@@ -453,7 +455,9 @@ &d0_gmac1 {
 	rst-gpios = <&porta 14 GPIO_ACTIVE_LOW>;
 	eswin,rgmiisel = <&pinctrl 0x294 0x3>;
 	eswin,led-cfgs = <0x6100 0xa40 0x420>;
-
+	dly-param-1000m = <0x25252525 0x80268025 0x26262626>;
+	dly-param-100m = <0x48484848 0x80588048 0x58585858>;
+	dly-param-10m = <0 0 0>;
 	status = "okay";
 };
 
diff --git a/arch/riscv/boot/dts/eswin/eic7702-evb-a1.dts b/arch/riscv/boot/dts/eswin/eic7702-evb-a1.dts
index 34502e9b98a1..058a28049e0f 100644
--- a/arch/riscv/boot/dts/eswin/eic7702-evb-a1.dts
+++ b/arch/riscv/boot/dts/eswin/eic7702-evb-a1.dts
@@ -775,7 +775,9 @@ &d0_gmac0 {
 	rst-gpios = <&porta 15 GPIO_ACTIVE_LOW>;
 	eswin,rgmiisel = <&pinctrl 0x290 0x3>;
 	eswin,led-cfgs = <0x6100 0xa40 0x420>;
-
+	dly-param-1000m = <0x23232323 0x800c8023 0x0c0c0c0c>;
+	dly-param-100m = <0x50505050 0x803f8050 0x3f3f3f3f>;
+	dly-param-10m = <0 0 0>;
 	status = "okay";
 };
 
@@ -785,7 +787,9 @@ &d1_gmac0 {
 	rst-gpios = <&d1_porta 13 GPIO_ACTIVE_LOW>;
 	eswin,rgmiisel = <&d1_pinctrl 0x290 0x3>;
 	eswin,led-cfgs = <0x6100 0xa40 0x420>;
-
+	dly-param-1000m = <0x23232323 0x800c8023 0x0c0c0c0c>;
+	dly-param-100m = <0x50505050 0x803f8050 0x3f3f3f3f>;
+	dly-param-10m = <0 0 0>;
 	status = "okay";
 };
 
@@ -795,7 +799,9 @@ &d0_gmac1 {
 	rst-gpios = <&porta 14 GPIO_ACTIVE_LOW>;
 	eswin,rgmiisel = <&pinctrl 0x294 0x3>;
 	eswin,led-cfgs = <0x6100 0xa40 0x420>;
-
+	dly-param-1000m = <0x25252525 0x80268025 0x26262626>;
+	dly-param-100m = <0x48484848 0x80588048 0x58585858>;
+	dly-param-10m = <0 0 0>;
 	status = "okay";
 };
 
@@ -805,7 +811,9 @@ &d1_gmac1 {
 	rst-gpios = <&d1_porta 14 GPIO_ACTIVE_LOW>;
 	eswin,rgmiisel = <&d1_pinctrl 0x294 0x3>;
 	eswin,led-cfgs = <0x6100 0xa40 0x420>;
-
+	dly-param-1000m = <0x25252525 0x80268025 0x26262626>;
+	dly-param-100m = <0x48484848 0x80588048 0x58585858>;
+	dly-param-10m = <0 0 0>;
 	status = "okay";
 };
 
diff --git a/arch/riscv/boot/dts/eswin/eswin-win2030-die0-soc.dtsi b/arch/riscv/boot/dts/eswin/eswin-win2030-die0-soc.dtsi
index e6d2005d854b..05e8a597d225 100644
--- a/arch/riscv/boot/dts/eswin/eswin-win2030-die0-soc.dtsi
+++ b/arch/riscv/boot/dts/eswin/eswin-win2030-die0-soc.dtsi
@@ -417,6 +417,7 @@ d0_gmac0: ethernet@50400000 {
 			dma-noncoherent;
 			eswin,hsp_sp_csr = <&d0_hsp_sp_csr 0x1030 0x100 0x108>;
 			eswin,syscrg_csr = <&d0_sys_crg 0x148 0x14c>;
+			eswin,dly_hsp_reg = <0x114 0x118 0x11c>;
 			snps,axi-config = <&d0_stmmac_axi_setup>;
 			d0_stmmac_axi_setup: stmmac-axi-config {
 				snps,blen = <0 0 0 0 16 8 4>;
@@ -447,6 +448,7 @@ d0_gmac1: ethernet@50410000 {
 			dma-noncoherent;
 			eswin,hsp_sp_csr = <&d0_hsp_sp_csr 0x1034 0x200 0x208>;
 			eswin,syscrg_csr = <&d0_sys_crg 0x148 0x14c>;
+			eswin,dly_hsp_reg = <0x214 0x218 0x21c>;
 			snps,axi-config = <&d0_stmmac_axi_setup_gmac1>;
 			d0_stmmac_axi_setup_gmac1: stmmac-axi-config {
 				snps,blen = <0 0 0 0 16 8 4>;
diff --git a/arch/riscv/boot/dts/eswin/eswin-win2030-die1-soc.dtsi b/arch/riscv/boot/dts/eswin/eswin-win2030-die1-soc.dtsi
index d216ce3c784c..51734f0ecfe1 100644
--- a/arch/riscv/boot/dts/eswin/eswin-win2030-die1-soc.dtsi
+++ b/arch/riscv/boot/dts/eswin/eswin-win2030-die1-soc.dtsi
@@ -1480,6 +1480,7 @@ d1_gmac0: ethernet@70400000 {
 			dma-noncoherent;
 			eswin,hsp_sp_csr = <&d1_hsp_sp_csr 0x1030 0x100 0x108>;
 			eswin,syscrg_csr = <&d1_sys_crg 0x148 0x14c>;
+			eswin,dly_hsp_reg = <0x114 0x118 0x11c>;
 			snps,axi-config = <&d1_stmmac_axi_setup>;
 			d1_stmmac_axi_setup: stmmac-axi-config {
 				snps,blen = <0 0 0 0 16 8 4>;
@@ -1510,6 +1511,7 @@ d1_gmac1: ethernet@70410000 {
 			dma-noncoherent;
 			eswin,hsp_sp_csr = <&d1_hsp_sp_csr 0x1034 0x200 0x208>;
 			eswin,syscrg_csr = <&d1_sys_crg 0x148 0x14c>;
+			eswin,dly_hsp_reg = <0x214 0x218 0x21c>;
 			snps,axi-config = <&d1_stmmac_axi_setup_gmac1>;
 			d1_stmmac_axi_setup_gmac1: stmmac-axi-config {
 				snps,blen = <0 0 0 0 16 8 4>;
diff --git a/drivers/net/ethernet/stmicro/stmmac/dwmac-win2030.c b/drivers/net/ethernet/stmicro/stmmac/dwmac-win2030.c
index e3b6c28394b6..6ae7b0ecfb86 100644
--- a/drivers/net/ethernet/stmicro/stmmac/dwmac-win2030.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac-win2030.c
@@ -1,6 +1,22 @@
-// SPDX-License-Identifier: GPL-2.0-only
+// SPDX-License-Identifier: GPL-2.0
 /*
  * Eswin DWC Ethernet linux driver
+ *
+ * Copyright 2024, Beijing ESWIN Computing Technology Co., Ltd.. All rights reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, version 2.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Authors: iangshuang <liangshuang@eswincomputing.com>
  */
 
 #include <linux/clk.h>
@@ -66,6 +82,10 @@ struct dwc_qos_priv {
 	struct gpio_desc *phy_reset;
 	struct stmmac_priv *stmpriv;
 	int phyled_cfgs[3];
+	unsigned int dly_hsp_reg[3];
+	unsigned int dly_param_1000m[3];
+	unsigned int dly_param_100m[3];
+	unsigned int dly_param_10m[3];
 };
 
 static int dwc_eth_dwmac_config_dt(struct platform_device *pdev,
@@ -198,21 +218,15 @@ static int eswin_eth_sid_cfg(struct device *dev)
 static void dwc_qos_fix_speed(void *priv, unsigned int speed, unsigned int mode)
 {
 	unsigned long rate = 125000000;
-	int err, data = 0;
+	int i, err, data = 0;
 	struct dwc_qos_priv *dwc_priv = (struct dwc_qos_priv *)priv;
 
 	switch (speed) {
 	case SPEED_1000:
 		rate = 125000000;
 
-		if ((dwc_priv->dev_id & 0x1) == 0) {
-			regmap_write(dwc_priv->hsp_regmap, 0x118, 0x800c8023);
-			regmap_write(dwc_priv->hsp_regmap, 0x11c, 0x0c0c0c0c);
-			regmap_write(dwc_priv->hsp_regmap, 0x114, 0x23232323);
-		} else {
-			regmap_write(dwc_priv->hsp_regmap, 0x218, 0x80268025);
-			regmap_write(dwc_priv->hsp_regmap, 0x21c, 0x26262626);
-			regmap_write(dwc_priv->hsp_regmap, 0x214, 0x25252525);
+		for (i = 0; i < 3; i++) {
+			regmap_write(dwc_priv->hsp_regmap, dwc_priv->dly_hsp_reg[i], dwc_priv->dly_param_1000m[i]);
 		}
 
 		if (dwc_priv->stmpriv) {
@@ -226,14 +240,8 @@ static void dwc_qos_fix_speed(void *priv, unsigned int speed, unsigned int mode)
 	case SPEED_100:
 		rate = 25000000;
 
-		if ((dwc_priv->dev_id & 0x1) == 0) {
-			regmap_write(dwc_priv->hsp_regmap, 0x118, 0x803f8050);
-			regmap_write(dwc_priv->hsp_regmap, 0x11c, 0x3f3f3f3f);
-			regmap_write(dwc_priv->hsp_regmap, 0x114, 0x50505050);
-		} else {
-			regmap_write(dwc_priv->hsp_regmap, 0x218, 0x80588048);
-			regmap_write(dwc_priv->hsp_regmap, 0x21c, 0x58585858);
-			regmap_write(dwc_priv->hsp_regmap, 0x214, 0x48484848);
+		for (i = 0; i < 3; i++) {
+			regmap_write(dwc_priv->hsp_regmap, dwc_priv->dly_hsp_reg[i], dwc_priv->dly_param_100m[i]);
 		}
 
 		if (dwc_priv->stmpriv) {
@@ -247,14 +255,8 @@ static void dwc_qos_fix_speed(void *priv, unsigned int speed, unsigned int mode)
 	case SPEED_10:
 		rate = 2500000;
 
-		if ((dwc_priv->dev_id & 0x1) == 0) {
-			regmap_write(dwc_priv->hsp_regmap, 0x118, 0x0);
-			regmap_write(dwc_priv->hsp_regmap, 0x11c, 0x0);
-			regmap_write(dwc_priv->hsp_regmap, 0x114, 0x0);
-		} else {
-			regmap_write(dwc_priv->hsp_regmap, 0x218, 0x0);
-			regmap_write(dwc_priv->hsp_regmap, 0x21c, 0x0);
-			regmap_write(dwc_priv->hsp_regmap, 0x214, 0x0);
+		for (i = 0; i < 3; i++) {
+			regmap_write(dwc_priv->hsp_regmap, dwc_priv->dly_hsp_reg[i], dwc_priv->dly_param_10m[i]);
 		}
 
 		if (dwc_priv->stmpriv) {
@@ -342,11 +344,11 @@ static int dwc_qos_probe(struct platform_device *pdev,
 	dwc_priv->dev = &pdev->dev;
 	dwc_priv->phy_reset = devm_gpiod_get(&pdev->dev, "rst", GPIOD_OUT_LOW);
 	if (IS_ERR(dwc_priv->phy_reset)) {
-		dev_err(&pdev->dev, "Reset gpio not specified\n");
-		return -EINVAL;
+		dev_info(&pdev->dev, "Reset gpio not specified\n");
 	}
 
-	gpiod_set_value(dwc_priv->phy_reset, 0);
+	if (dwc_priv->phy_reset)
+		gpiod_set_value(dwc_priv->phy_reset, 0);
 
 	dwc_priv->rgmii_sel = syscon_regmap_lookup_by_phandle(pdev->dev.of_node, "eswin,rgmiisel");
 	if (IS_ERR(dwc_priv->rgmii_sel)){
@@ -369,6 +371,30 @@ static int dwc_qos_probe(struct platform_device *pdev,
 		dev_warn(&pdev->dev, "can't get led cfgs for 10Mbps mode (%d)\n", ret);
 	}
 
+	ret = of_property_read_variable_u32_array(pdev->dev.of_node, "eswin,dly_hsp_reg", &dwc_priv->dly_hsp_reg[0], 3, 0);
+	if (ret != 3) {
+		dev_err(&pdev->dev, "can't get delay hsp reg.ret(%d)\n", ret);
+		return ret;
+	}
+
+	ret = of_property_read_variable_u32_array(pdev->dev.of_node, "dly-param-1000m", &dwc_priv->dly_param_1000m[0], 3, 0);
+	if (ret != 3) {
+		dev_err(&pdev->dev, "can't get delay param for 1Gbps mode (%d)\n", ret);
+		return ret;
+	}
+
+	ret = of_property_read_variable_u32_array(pdev->dev.of_node, "dly-param-100m", &dwc_priv->dly_param_100m[0], 3, 0);
+	if (ret != 3) {
+		dev_err(&pdev->dev, "can't get delay param for 100Mbps mode (%d)\n", ret);
+		return ret;
+	}
+
+	ret = of_property_read_variable_u32_array(pdev->dev.of_node, "dly-param-10m", &dwc_priv->dly_param_10m[0], 3, 0);
+	if (ret != 3) {
+		dev_err(&pdev->dev, "can't get delay param for 10Mbps mode (%d)\n", ret);
+		return ret;
+	}
+
 	ret = of_property_read_u32_index(pdev->dev.of_node, "eswin,rgmiisel", 1, &rgmiisel_offset);
 	if (ret) {
 		dev_err(&pdev->dev, "can't get rgmiisel_offset (%d)\n", ret);
@@ -487,7 +513,8 @@ static int dwc_qos_remove(struct platform_device *pdev)
 	reset_control_assert(dwc_priv->rst);
 	dwc_clks_config(dwc_priv, false);
 
-	devm_gpiod_put(&pdev->dev, dwc_priv->phy_reset);
+	if (dwc_priv->phy_reset)
+		devm_gpiod_put(&pdev->dev, dwc_priv->phy_reset);
 
 	return 0;
 }
-- 
2.49.0

